### ISSUE TITLE
sql/stats: forecast of zero rows causes bad plans

### ISSUE THREAD

**Issue by michae2** [cockroachdb/cockroach#119967]
[Statistics forecasting](https://www.cockroachlabs.com/docs/v23.2/cost-based-optimizer#enable-and-disable-forecasted-statistics-for-tables) was designed under the assumption that most row counts in a database grow over time. For counts that shrink over time, a forecast of zero rows can cause bad plans.

For example, say we have a large table, and are gradually filling in NULL values for an indexed column:

```sql
CREATE TABLE st (s int, t timestamptz, INDEX (s), INDEX (t));
INSERT INTO st SELECT generate_series(0, 99999), NULL;

-- do this in a loop
UPDATE st SET t = current_timestamp() WHERE s >= $1 AND s < $1 + 10 AND t IS NULL;
```

Assuming rows are updated at a constant rate, and statistics are automatically collected every 20% of row modifications, there should be a linear downward trend in the null row count for `t`. After we have updated 80% of the rows in the table, we'd expect the forecast to predict zero rows with `t IS NULL`. When this happens, the plan will switch from reading ~10 rows from the index on `s` to reading ~20k rows from the index on `t`.

We should be more cautious when forecasting a downward-trending row count. Some ideas we've discussed include:
- clamp predicted values to some constant greater than zero
- use a different regression model than linear regression
- force automatic stats collection near predicted inflection points
- decrease the forecast date until no values are predicted to be <= 0

Using the current time for the forecast time (instead of the predicted next auto stats collection time) would also be a fix.

Jira issue: CRDB-36418

**Comment by michae2** [comment]
I think both this and https://github.com/cockroachlabs/support/issues/2792 are strong arguments for picking something closer to the current time for the forecast, rather than the predicted time of the next automatic statistics collection. I've resisted doing this because it makes forecasts nondeterministic, but I think there are ways we could reduce the pain of nondeterminism:
1. round the current time down to the nearest hour / minute / some known interval, and use this for the forecast time, to get pseudo-determinism
2. for AOST queries use the calculated txn time instead of the current time
3. log changes in forecasts if they're significant, or plot them in some time-series chart
4. add a time to `SHOW STATISTICS WITH FORECAST` so that we can see what was happening at a certain time
5. add current forecasts to statement diagnostics bundles

**Comment by mgartner** [comment]
> round the current time down to the nearest hour / minute / some known interval, and use this for the forecast time, to get pseudo-determinism

Why is this required? As long as we know the timestamp being forecasted to, the forecast should be deterministic, correct?

> add current forecasts to statement diagnostics bundles

Alternatively, include the forecast timestamp in bundles and allow a session setting for setting forcing the forecast timestamp so we can reproduce the conditions under which the bundle was created. One downside of this is if stats from different tables can be forecasted at different times, then we'd need a way to set multiple forecast times when reproducing.

**Comment by michae2** [comment]
[triage] Rough plan is to backport a simple change setting minimums higher than 0 (2) and consider more involved change for 24.2 (forecasting closer to current time).

**Comment by michae2** [comment]
Here's a small repro on v23.2.10. First, set up an insecure demo cluster, with a table with 50k rows:

```
cockroach demo --insecure
CREATE TABLE ab (a PRIMARY KEY, b) AS SELECT generate_series(1, 50000), 1;
```

Then, run the following shell command, which will gradually delete all rows in the table over the course of about 5-10 minutes.

```zsh
for ((i=1; i<=50000; i += 20))
do
    ODD_KEYS="(${i}"
    for ((j=i+2; j<i+20; j += 2))
    do
        ODD_KEYS+=", ${j}"
    done
    ODD_KEYS+=")"
    EVEN_KEYS="(0"
    for ((j=i+1; j<i+20; j += 2))
    do
        EVEN_KEYS+=", ${j}"
    done
    EVEN_KEYS+=")"
    ./cockroach sql --insecure \
        -e "SET disallow_full_table_scans = on; SELECT * FROM ab WHERE a IN $ODD_KEYS" \
        -e "SET disallow_full_table_scans = on; DELETE FROM ab WHERE a IN $ODD_KEYS" \
        -e "SET disallow_full_table_scans = on; SELECT * FROM ab WHERE a IN $EVEN_KEYS" \
        -e "SET disallow_full_table_scans = on; DELETE FROM ab WHERE a IN $EVEN_KEYS" \
        > /dev/null
done
```

As the delete workload runs, keep an eye on table statistics for ab. Near the end of the workload, the forecast will show 0 rows in ab:

```
root@127.0.0.1:26257/defaultdb> SHOW STATISTICS FOR TABLE ab WITH FORECAST;
  statistics_name | column_names |            created            | row_count | distinct_count | null_count | avg_size | partial_predicate |    histogram_id     | full_histogram_id
------------------+--------------+-------------------------------+-----------+----------------+------------+----------+-------------------+---------------------+--------------------
  __auto__        | {a}          | 2024-09-13 19:14:24.803277+00 |     49900 |          49418 |          0 |        3 | NULL              | 1003305254278430721 |                 0
  __auto__        | {b}          | 2024-09-13 19:14:24.803277+00 |     49900 |              1 |          0 |        2 | NULL              | 1003305254281740289 |                 0
  __auto__        | {a}          | 2024-09-13 19:15:24.871233+00 |     43140 |          42692 |          0 |        3 | NULL              | 1003305451109449729 |                 0
  __auto__        | {b}          | 2024-09-13 19:15:24.871233+00 |     43140 |              1 |          0 |        2 | NULL              | 1003305451112923137 |                 0
  __auto__        | {a}          | 2024-09-13 19:18:24.965008+00 |     23160 |          22958 |          0 |        3 | NULL              | 1003306041240518657 |                 0
  __auto__        | {b}          | 2024-09-13 19:18:24.965008+00 |     23160 |              1 |          0 |        2 | NULL              | 1003306041243992065 |                 0
  __auto__        | {a}          | 2024-09-13 19:19:25.029276+00 |     16540 |          16427 |          0 |        3 | NULL              | 1003306238058823681 |                 0
  __auto__        | {b}          | 2024-09-13 19:19:25.029276+00 |     16540 |              1 |          0 |        2 | NULL              | 1003306238062231553 |                 0
  __auto__        | {a}          | 2024-09-13 19:20:25.123504+00 |      9900 |           9872 |          0 |        3 | NULL              | 1003306434971074561 |                 0
  __auto__        | {b}          | 2024-09-13 19:20:25.123504+00 |      9900 |              1 |          0 |        2 | NULL              | 1003306434974908417 |                 0
  __forecast__    | {a}          | 2024-09-13 19:21:55.203561+00 |         0 |              0 |          0 |        3 | NULL              |                   0 |              NULL
  __forecast__    | {b}          | 2024-09-13 19:21:55.203561+00 |         0 |              0 |          0 |        2 | NULL              |                   0 |              NULL
(12 rows)

Time: 2ms total (execution 2ms / network 0ms)
```

We can see in the SQL dashboard of the console that, at this point, full table scans start happening. This is because `large_full_scan_rows` is set to 1000 by default, so a full scan with a 0-row estimate is still allowed.

---

Now, to test the fix that is present in v23.2.10 but disabled, run the experiment again but with cluster setting `sql.stats.forecasts.max_decrease` set:

```
cockroach demo --insecure
CREATE TABLE ab (a PRIMARY KEY, b) AS SELECT generate_series(1, 50000), 1;
SET CLUSTER SETTING sql.stats.forecasts.max_decrease = 0.3;
```

At the end of the workload, the forecast should still be above 0:

```
root@127.0.0.1:26257/defaultdb> SELECT count(*) FROM ab;
  count
---------
      0
(1 row)

Time: 22ms total (execution 22ms / network 1ms)

root@127.0.0.1:26257/defaultdb> SHOW STATISTICS FOR TABLE ab WITH FORECAST;
  statistics_name | column_names |            created            | row_count | distinct_count | null_count | avg_size | partial_predicate |    histogram_id     | full_histogram_id
------------------+--------------+-------------------------------+-----------+----------------+------------+----------+-------------------+---------------------+--------------------
  __auto__        | {a}          | 2024-09-13 19:25:55.240838+00 |     36740 |          36212 |          0 |        3 | NULL              | 1003307516704882689 |                 0
  __auto__        | {b}          | 2024-09-13 19:25:55.240838+00 |     36740 |              1 |          0 |        2 | NULL              | 1003307516708585473 |                 0
  __auto__        | {a}          | 2024-09-13 19:27:55.308934+00 |     23300 |          23096 |          0 |        3 | NULL              | 1003307910143967233 |                 0
  __auto__        | {b}          | 2024-09-13 19:27:55.308934+00 |     23300 |              1 |          0 |        2 | NULL              | 1003307910147964929 |                 0
  __auto__        | {a}          | 2024-09-13 19:28:55.387703+00 |     16660 |          16543 |          0 |        3 | NULL              | 1003308107009556481 |                 0
  __auto__        | {b}          | 2024-09-13 19:28:55.387703+00 |     16660 |              1 |          0 |        2 | NULL              | 1003308107014438913 |                 0
  __auto__        | {a}          | 2024-09-13 19:29:55.466406+00 |     10020 |           9980 |          0 |        3 | NULL              | 1003308303875047425 |                 0
  __auto__        | {b}          | 2024-09-13 19:29:55.466406+00 |     10020 |              1 |          0 |        2 | NULL              | 1003308303879208961 |                 0
  __auto__        | {a}          | 2024-09-13 19:30:55.514651+00 |      3420 |           3418 |          0 |        3 | NULL              | 1003308500634501121 |                 0
  __auto__        | {b}          | 2024-09-13 19:30:55.514651+00 |      3420 |              1 |          0 |        2 | NULL              | 1003308500637974529 |                 0
  __forecast__    | {a}          | 2024-09-13 19:32:10.583104+00 |      1026 |           1025 |          0 |        3 | NULL              |                   0 |              NULL
  __forecast__    | {b}          | 2024-09-13 19:32:10.583104+00 |      1026 |              1 |          0 |        2 | NULL              |                   0 |              NULL
(12 rows)

Time: 8ms total (execution 8ms / network 1ms)
```

And there should have been no full table scans planned.

Eventually, after a stats collection measures 0 rows, the forecast will also show 0 rows. But by this point the table actually does have 0 rows:

```
root@127.0.0.1:26257/defaultdb> show statistics for table ab with forecast;
  statistics_name | column_names |            created            | row_count | distinct_count | null_count | avg_size | partial_predicate |    histogram_id     | full_histogram_id
------------------+--------------+-------------------------------+-----------+----------------+------------+----------+-------------------+---------------------+--------------------
  __auto__        | {a}          | 2024-09-13 19:27:55.308934+00 |     23300 |          23096 |          0 |        3 | NULL              | 1003307910143967233 |                 0
  __auto__        | {b}          | 2024-09-13 19:27:55.308934+00 |     23300 |              1 |          0 |        2 | NULL              | 1003307910147964929 |                 0
  __auto__        | {a}          | 2024-09-13 19:28:55.387703+00 |     16660 |          16543 |          0 |        3 | NULL              | 1003308107009556481 |                 0
  __auto__        | {b}          | 2024-09-13 19:28:55.387703+00 |     16660 |              1 |          0 |        2 | NULL              | 1003308107014438913 |                 0
  __auto__        | {a}          | 2024-09-13 19:29:55.466406+00 |     10020 |           9980 |          0 |        3 | NULL              | 1003308303875047425 |                 0
  __auto__        | {b}          | 2024-09-13 19:29:55.466406+00 |     10020 |              1 |          0 |        2 | NULL              | 1003308303879208961 |                 0
  __auto__        | {a}          | 2024-09-13 19:30:55.514651+00 |      3420 |           3418 |          0 |        3 | NULL              | 1003308500634501121 |                 0
  __auto__        | {b}          | 2024-09-13 19:30:55.514651+00 |      3420 |              1 |          0 |        2 | NULL              | 1003308500637974529 |                 0
  __auto__        | {a}          | 2024-09-13 19:31:55.590423+00 |         0 |              0 |          0 |        0 | NULL              | 1003308697492717569 |                 0
  __auto__        | {b}          | 2024-09-13 19:31:55.590423+00 |         0 |              0 |          0 |        0 | NULL              | 1003308697498419201 |                 0
  __forecast__    | {a}          | 2024-09-13 19:32:55.660795+00 |         0 |              0 |          0 |        3 | NULL              |                   0 |              NULL
  __forecast__    | {b}          | 2024-09-13 19:32:55.660795+00 |         0 |              0 |          0 |        2 | NULL              |                   0 |              NULL
(12 rows)

Time: 10ms total (execution 9ms / network 1ms)
```
